### PR TITLE
Fix bug for deseri type error that cause data error

### LINKED ISSUE #4851: TiFlash crashed when PageStorage V3 restore
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run ZTO workload, kill TiFlash process randomly.

### 2. What did you expect to see? (Required)
TiFlash can restart and the ZTO workload can be executed successfully after TiFlash restart.
### 3. What did you see instead (Required)
TiFlash crashed when PageStorage V3 restore.
`[2022/05/09 23:44:15.381 +08:00] [ERROR] [Exception.cpp:85] ["DB::EngineStoreApplyRes DB::HandleAdminRaftCmd(const DB::EngineStoreServerWrap *, DB::BaseBuffView, DB::BaseBuffView, DB::RaftCmdHeader):Code: 0, e.displayText() = DB::Exception: Invalid page id [page_id=90.2018] [resolve_id=90.2010], e.what() = DB::Exception, Stack trace:          0x1d082a3\tStackTrace::StackTrace() [tiflash+30442147]                 \tdbms/src/Common/StackTrace.cpp:23        0x1d058a6\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+30431398]                 \tdbms/src/Common/Exception.h:41        0x7a3db20\tDB::PS::V3::PageDirectory::getNormalPageId(DB::UInt128, std::__1::shared_ptr<DB::PS::V3::PageDirectorySnapshot> const&, bool) const [tiflash+128178976]                 \tdbms/src/Storages/Page/V3/PageDirectory.cpp:830        0x7a16f1a\tDB::PS::V3::PageStorageImpl::getNormalPageIdImpl(unsigned long, unsigned long, std::__1::shared_ptr<DB::PageStorageSnapshot>, bool) [tiflash+128020250]                 \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:76        0x77f2091\tDB::PageReader::getNormalPageId(unsigned long) const [tiflash+125771921]                 \tdbms/src/Storages/Page/PageStorage.h:372        0x785ff61\tDB::DM::ColumnFileBig::deserializeMetadata(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&) [tiflash+126222177]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp:75        0x7881f53\tDB::DM::deserializeSavedColumnFilesInV3Format(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&, unsigned long) [tiflash+126361427]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFile_V3.cpp:86        0x787901a\tDB::DM::deserializeSavedColumnFiles(DB::DM::DMContext&, DB::DM::RowKeyRange const&, DB::ReadBuffer&) [tiflash+126324762]                 \tdbms/src/Storages/DeltaMerge/ColumnFile/ColumnFilePersisted.cpp:123        0x788add1\tDB::DM::ColumnFilePersistedSet::restore(DB::DM::DMContext&, DB::DM::RowKeyRange const&, unsigned long) [tiflash+126397905]                 \tdbms/src/Storages/DeltaMerge/Delta/ColumnFilePersistedSet.cpp:112        0x78849ea\tDB::DM::DeltaValueSpace::restore(DB::DM::DMContext&, DB::DM::RowKeyRange const&, unsigned long) [tiflash+126372330]                 \tdbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp:59        0x77a549e\tDB::DM::Segment::restoreSegment(DB::DM::DMContext&, unsigned long) [tiflash+125457566]                 \tdbms/src/Storages/DeltaMerge/Segment.cpp:274        0x776ba0c\tDB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, long, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine> > const&, DB::DM::ColumnDefine const&, bool, unsigned long, DB::DM::DeltaMergeStore::Settings const&) [tiflash+125221388]                 \tdbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:263        0x76f706d\tDB::StorageDeltaMerge::getAndMaybeInitStore() [tiflash+124743789]                 \tdbms/src/Storages/StorageDeltaMerge.cpp:1527        0x76f9902\tDB::StorageDeltaMerge::flushCache(DB::Context const&, DB::DM::RowKeyRange const&) [tiflash+124754178]                 \tdbms/src/Storages/StorageDeltaMerge.cpp:773        0x78dc4e2\tDB::KVStore::tryFlushRegionCacheInStorage(DB::TMTContext&, DB::Region const&, Poco::Logger*) [tiflash+126731490]                 \tdbms/src/Storages/Transaction/KVStore.cpp:153        0x78deb5f\tDB::KVStore::handleUselessAdminRaftCmd(raft_cmdpb::AdminCmdType, unsigned long, unsigned long, unsigned long, DB::TMTContext&) [tiflash+126741343]                 \tdbms/src/Storages/Transaction/KVStore.cpp:384        0x78ded3c\tDB::KVStore::handleAdminRaftCmd(raft_cmdpb::AdminRequest&&, raft_cmdpb::AdminResponse&&, unsigned long, unsigned long, unsigned long, DB::TMTContext&) [tiflash+126741820]                 \tdbms/src/Storages/Transaction/KVStore.cpp:411        0x78f7685\tHandleAdminRaftCmd [tiflash+126842501]                 \tdbms/src/Storages/Transaction/ProxyFFI.cpp:114   0x7f4c3fb801de\traftstore::engine_store_ffi::_$LT$impl$u20$raftstore..engine_store_ffi..interfaces..root..DB..EngineStoreServerHelper$GT$::handle_admin_raft_cmd::hf504663a656e474e [libtiflash_proxy.so+21152222]   0x7f4c4031d669\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::exec_admin_cmd::hcd543b3e2a90c2c5 [libtiflash_proxy.so+29136489]   0x7f4c403109cf\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::hf45229906fb926c9 [libtiflash_proxy.so+29084111]   0x7f4c4030f305\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::ha4dbdc3a3458dcda [libtiflash_proxy.so+29078277]   0x7f4c403303cd\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h574db8b558a21550 [libtiflash_proxy.so+29213645]   0x7f4c403332a9\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::h5316fcd72343e9be [libtiflash_proxy.so+29225641]   0x7f4c40335cb0\t_$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::hec3475cd3a78b2fd [libtiflash_proxy.so+29236400]   0x7f4c3ffa2365\tbatch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h5af46678afa583e3 [libtiflash_proxy.so+25486181]   0x7f4c406689ab\tstd::sys_common::backtrace::__rust_begin_short_backtrace::h0b8b7288f3f90d60 [libtiflash_proxy.so+32590251]   0x7f4c3fe8e62c\tcore::ops::function::FnOnce::call_once$u7b$$
`
### 4. What is your TiFlash version? (Required)
8781b0fcb3394ebd7817fd1ddab713021076d242 or 8781b0fcb3394ebd7817fd1ddab713021076d242 with PR https://github.com/pingcap/tiflash/pull/4850

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: close #4851 

Problem Summary:
```
[2022/05/10 15:18:28.483 +08:00] [TRACE] [PageDirectoryFactory.cpp:131] ["PageDirectoryFactory:restoring record [type=1] [page_id=192.352] [ver=<3405,0>] [entry=PageEntry{file: 0, offset: 0x0, size: 0, checksum: 0x0}]"] [thread_id=1]
[2022/05/10 15:18:28.483 +08:00] [TRACE] [PageDirectoryFactory.cpp:131] ["PageDirectoryFactory:restoring record [type=8] [page_id=192.352] [ver=<3407,0>] [entry=PageEntry{file: 0, offset: 0x0, size: 0, checksum: 0x0}]"] [thread_id=1]
[2022/05/10 15:18:28.483 +08:00] [TRACE] [PageDirectoryFactory.cpp:131] ["PageDirectoryFactory:restoring record [type=6] [page_id=192.357] [ver=<3421,0>] [entry=PageEntry{file: 0, offset: 0x0, size: 0, checksum: 0x0}]"] [thread_id=1]
```

- The `page_id=192.352` type is `PUT_EXTERNAL(1)`, but it should be `VAR_EXTERNAL(7)`. Because this record has been dumped into `log_5_1`
- Then after restored, `PageDirectoryFactory::loadEdit` will deal this edit as `PUT_EXTERNAL(1)` following by `DEL(8)`, it will make that this page_id  `being_ref_count` be 0. MVCC will clean it in memory.

The main reason why "`page_id=192.352` type is `1`" is that we should not decode `VAR_EXTERNAL(7)`  to `PUT_EXTERNAL(1)` in `deserializePutExternalFrom`.

### What is changed and how it works?

- Changed `deserializePutExternalFrom `
- Added a log to check the similar situations.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/WAL/serialize.cpp [hunk 1]
```diff
@@ -152,7 +152,7 @@ void deserializePutExternalFrom([[maybe_unused]] const EditRecordType record_typ
     assert(record_type == EditRecordType::PUT_EXTERNAL || record_type == EditRecordType::VAR_EXTERNAL);
 
     PageEntriesEdit::EditRecord rec;
-    rec.type = EditRecordType::PUT_EXTERNAL;
+    rec.type = record_type;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/WAL/serialize.cpp:155]: also remove this line
https://github.com/pingcap/tiflash/blob/2cb02ae11882b8fcaf768f38642d53b113f20995/dbms/src/Storages/Page/V3/WAL/serialize.cpp#L103

Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/WAL/serialize.cpp:155]: Add related ut for VAR_XXX in `WALSeriTest` in `dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp`

Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/WAL/serialize.cpp:154]: done

File: dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp [hunk 2]
```diff
@@ -132,6 +132,33 @@ TEST(WALSeriTest, Upserts)
...
+    edit.varExternal(1, ver1_0, 2);
+    edit.varDel(1, ver2_0);
+    edit.varRef(2, ver3_0, 1);
+
+    auto deseri_edit = DB::PS::V3::ser::deserializeFrom(DB::PS::V3::ser::serializeTo(edit));
+    ASSERT_EQ(deseri_edit.size(), 3);
+    auto iter = deseri_edit.getRecords().begin();
+    EXPECT_EQ(iter->type, EditRecordType::VAR_EXTERNAL);
+    EXPECT_EQ(iter->page_id.low, 1);
+    EXPECT_EQ(iter->version, ver1_0);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp:151]: check the being_ref_count for VAR_EXTERNAL

And add test for VAR_ENTRY and also check the being_ref_count

Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp]: added.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by hehechen — APPROVED** [review]

**Pull request merged**